# Explanation of variant properties


| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| input.fields.InbreedingCoeff  | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |

```
{  
   "custom_annotations":{  
      "input":[  
         {  
            "fields":{  
               "InbreedingCoeff":0.0552,
               "SOR":0.829,
               "AF":0.651,
               "culprit":"FS",
               "AC":6339,
               "AN":9740,
               "FS":1.189,
               "ClippingRankSum":0.03,
               "MLEAC":6341,
               "ReadPosRankSum":0.326,
               "VQSLOD":3.37,
               "BaseQRankSum":0.481,
               "MLEAF":0.651,
               "MQ0":0,
               "QD":20.93,
               "ExcessHet":0,
               "MQRankSum":0.032,
               "MQ":23.88,
               "DP":245471
            },
            "allele":"G",
            "name":"22-38212762-A-G"
         }
      ],
      "gnomad_genomes_coding_autosomes":[  
         {  
            "name":"rs13911",
            "fields":{  
               "AF_NFE":0.64875,
               "AC_AFR":7224,
               "AN_Female":13822,
               "AC_NFE":9704,
               "Hom_FIN":722,
               "AF_AFR":0.828821,
               "AC_Male":12193,
               "AF_EAS":0.891708,
               "AN_Male":17078,
               "AC_ASJ":184,
               "AF_Female":0.709159,
               "AC_FIN":2260,
               "Hom":7974,
               "AN_NFE":14958,
               "AC_AMR":540,
               "AN_OTH":980,
               "Hom_NFE":3177,
               "AC_raw":22041,
               "Hom_AMR":171,
               "Hom_EAS":645,
               "AF_AMR":0.644391,
               "GC":1429,
               "AF_OTH":0.655102,
               "AF_Male":0.713959,
               "AN_AFR":8716,
               "AN_FIN":3490,
               "AN_EAS":1616,
               "AC_EAS":1441,
               "Hom_raw":7990,
               "GC_raw":1445,
               "AC_OTH":642,
               "Hom_Female":3561,
               "Hom_OTH":215,
               "AF_ASJ":0.609272,
               "AN_ASJ":302,
               "AF_raw":0.711184,
               "Hom_ASJ":55,
               "AN_raw":30992,
               "AN_AMR":838,
               "AC_Female":9802,
               "Hom_AFR":2989,
               "Hom_Male":4413,
               "AF_FIN":0.647564
            },
            "allele":"G"
         }
      ],
      "dbsnp":[  
         {  
            "fields":{  
               "VLD":null,
               "VC":"SNV",
               "U3":null,
               "CAF":0.2462,
               "WGT":1,
               "GENEINFO":"GCAT:23464",
               "VP":"0x05010088000517053f000100",
               "SSR":0,
               "SAO":0
            },
            "allele":"G",
            "name":"rs13911"
         }
      ],
      "gnomad_exomes":[  
         {  
            "fields":{  
               "AN_AFR":14770,
               "AN_FIN":21580,
               "AF_Male":0.678768,
               "AF_OTH":0.642963,
               "GC":13029,
               "Hom_SAS":8220,
               "AF_AMR":0.63848,
               "Hom_AMR":6841,
               "AC_raw":167279,
               "Hom_EAS":6643,
               "Hom_NFE":22584,
               "AN_OTH":5400,
               "AC_AMR":21338,
               "Hom":56568,
               "AN_NFE":108778,
               "AC_FIN":13756,
               "AN_SAS":30712,
               "AF_Female":0.682046,
               "AF_EAS":0.879077,
               "AC_ASJ":5851,
               "AN_Male":132708,
               "AC_SAS":22369,
               "AF_SAS":0.728347,
               "AC_NFE":70249,
               "AN_Female":108846,
               "AF_AFR":0.825254,
               "Hom_FIN":4361,
               "AC_Male":90078,
               "AF_NFE":0.645802,
               "AC_AFR":12189,
               "Hom_Male":30978,
               "AF_FIN":0.637442,
               "AN_AMR":33420,
               "AC_Female":74238,
               "Hom_AFR":5020,
               "AN_raw":246190,
               "AF_raw":0.679471,
               "Hom_ASJ":1772,
               "AN_ASJ":9726,
               "AF_ASJ":0.601583,
               "Hom_OTH":1127,
               "AC_OTH":3472,
               "Hom_Female":25590,
               "GC_raw":13510,
               "Hom_raw":57694,
               "AC_EAS":15092,
               "AN_EAS":17168
            },
            "allele":"G",
            "name":"rs13911"
         }
      ],
      "kaviar":[  
         {  
            "fields":{  
               "AF":0.6665359,
               "AC":103649,
               "AN":155504
            },
            "allele":"G",
            "name":"rs13911"
         }
      ],
      "gnomad_genomes":[  
         {  
            "name":"rs13911",
            "fields":{  },
            "allele":"G"
         }
      ]
   },
   "transcript_consequences":[  
      {  
         "exac_af_afr":0.824,
         "exac_af_nfe":0.643,
         "exac_af_eas":0.878,
         "cadd_raw":-0.123251,
         "exac_af_popmax":0.878,
         "strand":1,
         "transcript_id":"ENST00000323205",
         "exac_af_female":0.692,
         "variant_allele":"G",
         "gene_symbol_source":"HGNC",
         "exacpli":0,
         "impact":"MODIFIER",
         "uniparc":[  
            "UPI000020770F"
         ],
         "gene_symbol":"GCAT",
         "exac_af_oth":0.713,
         "protein_id":"ENSP00000371110",
         "biotype":"protein_coding",
         "exac_af_sas":0.728,
         "consequence_terms":[  
            "intron_variant"
         ],
         "intron":"9/9",
         "exac_af_fin":0.634,
         "exac_af_adj":0.685,
         "hgnc_id":4188,
         "ccds":"CCDS54527.1",
         "exac_af_male":0.679,
         "swissprot":[  
            "O75600"
         ],
         "cadd_phred":1.563,
         "canonical":1,
         "exac_af":0.682,
         "exac_af_consanguineous":0.73,
         "exac_af_amr":0.624,
         "hgvsc":"ENST00000323205.6:c.*7+30A>G",
         "gene_id":"ENSG00000100116"
      }
   ],
   "end":38212762,
   "input":"22\t38212762\t22-38212762-A-G\tA\tG\t4821151.27\tPASS\tAC=6339;AF=0.651;AN=9740;BaseQRankSum=0.481;ClippingRankSum=0.030;DB;DP=245471;ExcessHet=0.0000;FS=1.189;InbreedingCoeff=0.0552;MLEAC=6341;MLEAF=0.651;MQ=23.88;MQ0=0;MQRankSum=0.032;POSITIVE_TRAIN_SITE;QD=20.93;ReadPosRankSum=0.326;SOR=0.829;VQSLOD=3.37;culprit=FS",
   "seq_region_name":"22",
   "allele_string":"A/G",
   "start":38212762,
   "most_severe_consequence":"intron_variant",
   "strand":1,
   "variant_class":"SNV",
   "assembly_name":"GRCh37",
   "colocated_variants":[  
      {  
         "ea_allele":"G",
         "exac_sas_maf":0.7281,
         "exac_oth_maf":0.7132,
         "sas_allele":"G",
         "exac_maf":0.682,
         "amr_allele":"G",
         "exac_amr_maf":0.6236,
         "end":38212762,
         "aa_maf":0.8224,
         "minor_allele":"A",
         "allele_string":"A/G",
         "exac_oth_allele":"G",
         "exac_eas_maf":0.8778,
         "afr_maf":0.8457,
         "sas_maf":0.7147,
         "exac_afr_allele":"G",
         "minor_allele_freq":0.2462,
         "eur_maf":0.6392,
         "ea_maf":0.6436,
         "exac_adj_allele":"G",
         "exac_adj_maf":0.6849,
         "eas_allele":"G",
         "eur_allele":"G",
         "eas_maf":0.8472,
         "afr_allele":"G",
         "exac_sas_allele":"G",
         "id":"rs13911",
         "amr_maf":0.6643,
         "exac_afr_maf":0.8238,
         "start":38212762,
         "strand":1,
         "exac_nfe_allele":"G",
         "exac_amr_allele":"G",
         "aa_allele":"G",
         "exac_fin_maf":0.634,
         "exac_fin_allele":"G",
         "seq_region_name":22,
         "exac_nfe_maf":0.6428,
         "exac_allele":"G",
         "exac_eas_allele":"G"
      }
   ],
   "id":"22-38212762-A-G"
}
```


## Variant quality

### FILTER
Values:
* PASS
* FAIL
* VQSR

### FS: Fisher Strand
Whether there is a bias in the fisher strand.

### BaseQRankSum

### MQ
Mapping quality

### QD
Quality depth

### ClippingRankSum

### MQRankSum

## Variant Frequency

### Kaviar

The [Kaviar db](db.systemsbiology.net/kaviar/)

Released: February 29, 2016 (version 160204-Public)
Kaviar contains 162 million SNV sites (including 25M not in dbSNP) and incorporates data from 35 projects encompassing 77,781 individuals (13.2K whole genome, 64.6K exome).
* Kaviar also contains 50 million short indels and substitutions from a subset of the data sources.
* Kaviar excludes cancer genomes but includes some data from cell lines and individuals affected by disease.
* An effort was made to exclude data from related individuals.

### Gnomad

The Genome Aggregation Database (gnomAD) is a resource developed by an international coalition of investigators, with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects, and making summary data available for the wider scientific community.

The data set provided on this website spans 123,136 exome sequences and 15,496 whole-genome sequences from unrelated individuals sequenced as part of various disease-specific and population genetic studies.


### ExAC

## Variant function impact

### CADD
The CADD is not transcript specific but depends on the DNA

### Transcript Consequences

The trancript consequences field contains the functional impact of the variant on all transcript on which it occurs.
Each transcript consequence contains:
* exon affected : the number of the exon affected
* hgvsc : change in cDNA (cDNA has introns spliced out)
* hgvsp: if the transcript is protein coding, what is the effect on the protein?

