{% extends "material_layout.html" %} {% block body %}
<div class="section">
    <div class="container">
        <h4 class="center" style="padding-bottom:1rem;">About Phenopolis</h4>
        <p style="font-size:1.2rem;">Phenopolis is an open platform for harmonization and analysis of sequencing and phenotype data. The Phenopolis code is available at github. Funding is currently provided mainly by RP Fighting Blindness, NIHR (UCL-Moorfields Eyes Hospital Biomedical Research Center) and in part by the UK Medical Research Council and the British Heart Foundation.</p>
        <p style="font-size:1.2rem;">The data set provided on this website spans <span class="number_samples">{{ total_patients }}</span> sequenced individuals. All of the raw data from these projects have been reprocessed through the same pipeline, and jointly variant-called to increase consistency across projects.</p>
        <p style="font-size:1.2rem;">Behind Phenopolis is a coalition of clinicians, geneticists, bioinformaticians and software engineers seeking to aggregate and harmonize sequencing and phenotype data using the Human Phenotype Ontology.</p>
    </div>
</div>
<div class="section" id="team" style="background-color: #ececec;">
    <div class="container">
        <h4 class="center" style="padding-bottom:1rem;">The Team behind Phenopolis</h4>
        <div class="row">
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/nikolas_pic.jpeg" alt="" class="circle">
                        <h5>Dr. Nikolas Pontikos</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;UCL Genetics Institute;</span> 
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Institute of Ophthalmology, UCL;</span>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp; Moorfields Eye Hospital, London</span>
                        <p>I started the development of Phenopolis.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/jing_pic.jpeg" alt="" class="circle">
                        <h5>Dr. Jing Yu</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Nuffield Department of Clinical Neurosciences, University of Oxford</span> 
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;John Radcliffe Hospital, Oxford</span> 

                        <p>I helped construct the website, and developed pubmedBatch, Phenogenon and several other analyses / visualisations in Phenopolis.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
        </div>

        <div class="row">
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/cian_pic.jpeg" alt="" class="circle">
                        <h5>Dr. Cian Murphy</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;UCL Genetics Institute</span> 
                        <p>I developed gene burden tests and run the CNV pipeline (data to be integrated).</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/daniel_pic.jpeg" alt="" class="circle">
                        <h5>Daniel Greene</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Department of Haematology, University of Cambridge;</span> 
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Medical Research Council Biostatistics Unit, Cambridge</span>                         
                        <p>I helped integrate the SimReg method for discovering genotype-phenotype associations into phenopolis. SimReg evaluates the evidence of association between ontologically encoded phenotypes and rare alleles in given genes. Where there is strong evidence of association, SimReg also estimates a characteristic phenotype for the disease caused by pathogenic variants.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
        </div>

        <div class="row">
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/ismail_pic.jpeg" alt="" class="circle">
                        <h5>Ismail Moghul</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;UCL Cancer Institute</span> 
                        <p>I worked on the front-end aspect of Phenopolis with a particular focus on improving the user interphase and user experience as well as the general performance of the web application.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/lucy_pic.jpeg" alt="" class="circle">
                        <h5>Dr. Lucy Withington</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;RP Fighting Blindness</span> 
                        <p>I wrote unit tests and did security and usability tests on the website.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
        </div>

        <div class="row">
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/tom_pic.jpeg" alt="" class="circle">
                        <h5>Dr. Tom Vulliamy</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Blizard Institute, Queen Mary University of London</span> 
                        <p>I have been sequencing patients who have a bone marrow failure syndrome that is clinically and genetically heterogeneous. <br>
                        We have been trying to identify the genes responsible. Although we have had some success, a significant proportion remain uncharacterised, and we are interesting in novel analysis approaches using the Human Phenotype Ontology.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/fiona_pic.png" alt="" class="circle">
                        <h5>Dr. Fiona Blanco-Kelly</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;Institute of Ophthalmology, UCL;</span>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp; Moorfields Eye Hospital, London</span>
                        <p>I entered and curated a large number of phenotypes encoded using the Human Phenotype Ontology</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
        </div>

        <div class="row">
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/baron_pic.jpeg" alt="" class="circle">
                        <h5>Baron Koylass</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;UCL Genetics Institute</span> 
                        <p>I am an MSc student of the Genetics of Human Disease program at University College London, doing my Masters project on Phenopolis.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
            <div class="col l6 m12">
                <ul class="collection">
                    <li class="collection-item avatar">
                        <img src="/static/images/team/sajid_pic.jpeg" alt="" class="circle">
                        <h5>Sajid Mughal</h5>
                        <span class="blue-text"><i class="material-icons">school</i>&nbsp;UCL Genetics Institute</span> 
                        <p>I am a software developer developing graph databases for Phenopolis. My project page is <a href="https://github.com/sajid-mughal/pheno4j">here</a>.</p>
                    </li>
                    <div class="clearfix"></div>
                </ul>
            </div>
        </div>
    </div>
</div>
<div class="section">
    <div class="container">
        <h4 class="center" style="padding-bottom:1rem;">The Pitch</h4>
<p style="font-size:1.2rem;">
Rare disease diagnostic research labs and large initiatives such as Genomics England are generating growing amounts of next generation sequencing data and standardised clinical phenotypes encoded with the Human Phenotype Ontology (HPO).
</p>

<p style="font-size:1.2rem;">
There is a clear need for a free, easy and effective method of analysing and sharing this wealth of information between clinicians and researchers.
</p>

<p style="font-size:1.2rem;">
With this goal in mind, we have developed, Phenopolis, an open-source research and diagnosis tool. While there are already several commercial tools that provide filtering and prioritisation of genetic variants (e.g Congenica, Saphetor, Omicia) in patient, there are currently only two other open-source alternatives (Gemini and Seqr), and we are not aware of any tools which supports the type of HPO analysis our platform supports.
</p>

<p style="font-size:1.2rem;">
Our software unifies existing HPO analysis methods (Exomiser, SimReg) but also our own novel analysis and visualisation methods (PubmedScore, Phenogenon), through an easily extensible interface. 
</p>

<p style="font-size:1.2rem;">
Phenopolis supports three visualisations of the data depending on the research question:
<ul>
    <li> A view of all the variants in a given patient and tools to filter and prioritise the variants to identify those that might be causing the phenotype.</li>
    <li> A view of all variants and phenotypes of the carriers in a gene to understand the phenotype signature linked to a gene. This is using our novel tool, Phenogenon, to give an overview of the HPO signature of a gene based on real-time filtering of genetic variants on frequency and consequence score.</li>
    <li> A phenotype view to understand the gene and gene modifiers that may be associated to a phenotype using gene burden tests and Phenogenon.</li>
</ul>
</p>

<p style="font-size:1.2rem;">
Our software is being actively used by UK universities, UCL, Queen Mary, Leeds, Oxford and Manchester, and has already been referenced in “The Human Phenotype Ontology 2017”  update published in the Nucleic Acid Research 2016 (<a href=https://academic.oup.com/nar/article/45/D1/D865/2574174/The-Human-Phenotype-Ontology-in-2017>doi:10.1093/nar/gkw1039</a>).
</p>

<p style="font-size:1.2rem;">
The team behind Phenopolis is a unique combination of software developers, PhD students, postdoctoral bioinformaticians, geneticists and clinicians, all dedicated to free software and data sharing, and has managed, in the space of a year with no project funding to make significant progress.
</p>

<p style="font-size:1.2rem;">
Seed funding would allow us to purchase the hardware to make our website reliable, add functionality and scale to provide a free service to anyone who wants to have their data analysed. After the funding period, Phenopolis will be connected to the Global Alliance for Genomics and Health network (which includes large patient database such  as DECIPHER, PhenomeCentral) and could be used by initiatives such as Genomics England.
</p>

<p style="font-size:1.2rem;">
Our mission is to guarantee that anyone around the world will be able to use our platform for free to address important research and clinical questions . We think that Phenopolis is promoting an extremely important message of openness and data sharing to solve human disease.
</p>
    </div>
</div>

{% endblock %}
