{% extends "layout.html" %}
{% block body %}
    <script>
        $('#navbar_form_container').hide();
    </script>
    <style>
        svg path {
            -webkit-transition: fill 0.8s; /* Safari */
            transition: fill 0.8s;
        }
        .popover {
            text-align: center;
        }
        .node:hover {
            cursor:pointer;
        }
        div.hpo {
            height:700px;
            width: 2000px;
        }

        h1 {
              font-family: "Lucida Console";
        }


    </style>
<div class="tabs">
<ul class="nav nav-tabs" role="tablists">
    <li class="active"><a href="#tab1" role="tab" data-toggle="tab">Search Page</a></li>
    <li><a href="#tab2" role="tab" data-toggle="tab">HPO Browser</a></li>
    {% if session["user"]!="demo" %}
        <li><a href="#tab3" role="tab" data-toggle="tab">Stats</a></li>
    {% endif %}
    <li><a href="#tab4" role="tab" data-toggle="tab">Publications</a></li>
    <li><a href="#tab5" role="tab" data-toggle="tab">Terms of Use</a></li>
</ul>

<div class="container">
    <h1 id="home-title">Phenopolis </h1>
  <div class="tab-content">

            <div id="tab1" class="tab-pane active">
                <div class="row">
                <div class="col-md-12">
                    <div id="home-searchbox">
                        <form action="/awesome">
                            <input type="submit" style="display: none;"/>
                            <input id="home-searchbox-input" name="query" class="form-control input-lg awesomebar" type="text" placeholder="Search for a phenotype, patient, gene, variant or region"/>
                            <input type="submit" style="position: absolute; left: -9999px"/>
                        </form>
                        <p class="text-muted">
                            Examples - Phenotype: <a href="/hpo/Nystagmus">Nystagmus</a>,
                            HPO: <a href="/hpo/HP:0000639">HP:0000639</a>,
                            Individual: <a href="/individual/WebsterURMD_Sample_GV4344">WebsterURMD_Sample_GV4344</a>,
                            Gene: <a href="/gene/DRAM2">DRAM2</a>,
                            Transcript: <a href="/transcript/ENST00000553765">ENST00000553765</a>,
                            Variant: <a href="/variant/22-38212762-A-G">22-38212762-A-G</a>
                        </p>
                    </div>
                </div>
                </div>
                <div class="row" style="font-size: 16px;">
                    <div class="col-md-8">
                        <h3 align="center"> <a href="/about" style="color: black;"> About  </a></h3>
                        <div>
                            <p>
                            <font face="Lucida Console">Phenopolis</font> is an open platform for harmonization
                            and analysis of sequencing and phenotype data.
                            The <font face="Lucida Console">Phenopolis</font> code is available at <a href=https://github.com/pontikos/phenopolis>github</a>.
                            Funding is currently provided mainly by <a href=http://www.rpfightingblindness.org.uk/home.php?home=yes> RP Fighting Blindness </a>, 
                            NIHR (UCL-Moorfields Eyes Hospital Biomedical Research Center)
                            and in part by the UK Medical Research Council and the British Heart Foundation.
                            </p>
							
                            <p>
                            {% if version_number %}
                            Version {{ version_number }}
                            {% endif %}
                            </p>

                            <h3 align="center"> Samples </h3>
                            <p>
                            Phenopolis currently contains {{ total_patients }} exomes ({{ male_patients }} male,
                            {{ female_patients }} female, and {{ unknown_patients }} unknown).
                            {% if session["user"]!="demo" %}
                                An interactive PCA plot of the patients ancestry is available <a href="/static/plots/UCLex_ancestry_pca.html"> here</a>.
                            {% endif %}
                            </p>

                            <h3 align="center"> Variants </h3>
                            <p>
                            There is a total of {{ '{0:,}'.format(total_variants) }} variants.
                            {{ '{0:,}'.format(pass_variants) }} variants pass the GATK filter.
                            </p>
                        </div>
                    </div>
                    {% if session["user"]!="demo" %}
                    <div class="col-md-4">
                        <h3 align="center">Recent News</h3>
                        <h4>December 1, 2015</h4> - First launch.
                        <h4>February 8, 2016</h4> - Demo at UKIRDC Oxford meeting
                        <h4>April 19, 2016</h4> - Demo at the UKIRDC Leeds meeting.
                        <h4>April 26, 2016</h4> - Demo at HPO day at UCL.
                        <h4>June 22, 2016</h4> - Demo at Curating the Clinical Genome.
                    </div>
                    {% endif %}
                </div>
             </div>

             <div id="tab2" class="tab-pane">
                 <div class='row'>
                     <div class='col-md-7'>
                        <div class='row'>
                            <label class='tip' title='individuals relatedness is calculated using KINSHIP'>
                              <input id='remove-related' type="checkbox" checked> Remove relatedness
                            </label>&nbsp; [<span id='num_of_patients'></span> patients]
                        </div>
                        <div class='row'>
                                <br/>
                                <form role="form" id="choice-form">
                                    <label class="radio-inline">
                                        <input type="radio" name="optradio" value="cohort" >Break down on cohort
                                    </label>
                                    <label class="radio-inline">
                                        <input type="radio" name="optradio" value="sex" checked>Break down on sex
                                    </label>
                                    <label class="radio-inline">
                                        <input type="radio" name="optradio" value="solved">Break down on solved status
                                    </label>
                                </form>
                        </div>
                        <div class='row'>
                            <div id='hpo-legend'></div>
                        </div>
                        <div class='row'>
                            <div class='hpo' id='hpo-dot'> </div>
                        </div>
                     </div><!--col-md-7-->
                     <div class='col-md-5' id='side-pane'><svg></svg></div>
                 </div><!--top row in tab2-->
             </div>

             {% if session["user"]!="demo" %}
             <div id="tab3" class="tab-pane">
                 <div>
                    {{image|safe}}
                 </div>
                 <table>
                     <thead> <tr> <th> Variants </th> </tr> </thead>
                     <tbody>
                     <tr> <td> all variants </td> <td> {{ '{0:,}'.format(total_variants) }} </td> </tr>
                     <tr> <td> pass variants </td> <td> {{ '{0:,}'.format(pass_variants) }} </td> </tr>
                     <tr> <td> non-pass variants </td> <td> {{'{0:,}'.format(nonpass_variants)}} </td> </tr>
                     </tbody>
                 </table>
                 <table>
                     <thead> <tr> <th> Patients </th> </tr> </thead>
                     <tbody>
                     <tr> <td> all patients </td> <td> {{ '{0:,}'.format(total_patients) }} </td> </tr>
                     </tbody>
                 </table>
            </div>
            {% endif %}

             <div id="tab4" class="tab-pane">

                 <b>Primary immune deficiencies:</b>

 <p>
Phosphoinositide 3-Kinase δ Gene Mutation Predisposes to Respiratory Infection and Airway Damage, Angulo I, Vadas O, Garçon F, Banham-Hall E, Plagnol V, Leahy TR, Baxendale H, Coulter T, Curtis J, Wu C, Blake-Palmer K, Perisic O, Smyth D, Maes M,Fiddler C, Juss J, Cilliers D, Markelj G, Chandra A, Farmer G, Kielkowska A, Clark J, Kracker S, Debré M, Picard C, Pellier I, Jabado N, Morris JA, Barcenas-Morales G, Fischer A, Stephens L, Hawkins P, Barrett JC, Abinun M, Clatworthy M, Durandy A, Doffinger R, Chilvers E, Cant AJ, Kumararatne D, Okkenhaug K,Williams RL, Condliffe A, Nejentsev S., Science 2013
</p>

<p>
LRBA gene deletion in a patient presenting with autoimmunity without hypogammaglobulinemia. Burns SO, Zenner HL, Plagnol V, Curtis J, Mok K, Eisenhut M, Kumararatne D, Doffinger R, Thrasher AJ, Nejentsev S.  J Allergy Clin Immunol. 2012
</p>


<b>Ophthalmology:</b>

<p>
Mutations in ARL2BP, Encoding ADP-Ribosylation-Factor-Like 2 Binding Protein, Cause Autosomal-Recessive Retinitis Pigmentosa, Davidson AE, Schwarz N, Zelinger L, Stern-Schneider G, Shoemark A, Spitzbarth B, Gross M, Laxer U, Sosna J, Sergouniotis PI, Waseem NH, Wilson R, Kahn RA, Plagnol V, Wolfrum U, Banin E, Hardcastle AJ, Cheetham ME, Sharon D, Webster AR, Am J Hum Genet. 2013
</p>

<p>
RP1L1 variants are associated with a spectrum of inherited retinal diseases including retinitis pigmentosa and occult macular dystrophy., Davidson AE, Sergouniotis PI, Mackay DS, Wright GA, Waseem NH, Michaelides M, Holder GE, Robson AG, Moore AT, Plagnol V, Webster AR. Hum Mutat. 2013
</p>

<p>
Recessive mutations in KCNJ13, encoding an inwardly rectifying potassium channel subunit, cause leber congenital amaurosis. Sergouniotis PI, Davidson AE, Mackay DS, Li Z, Yang X, Plagnol V, Moore AT, Webster AR. Am J Hum Genet. 2011
</p>

<p>
Biallelic mutations in PLA2G5, encoding group V phospholipase A2, cause
benign fleck retina., Sergouniotis PI, Davidson AE, Mackay DS, Lenassi E, Li Z, Robson AG, Yang X, Kam JH, Isaacs TW, Holder GE, Jeffery G, Beck JA, Moore AT, Plagnol V, Webster AR. Am J Hum Genet. 2011
</p>

<b>Dermatology:</b>

<p>
Mutations in AQP5, Encoding a Water-Channel Protein, Cause Autosomal-Dominant Diffuse Nonepidermolytic Palmoplantar Keratoderma.
Blaydon DC, Lind LK, Plagnol V, Linton KJ, Smith FJ, Wilson NJ, McLean WH, Munro CS, South AP, Leigh IM, O'Toole EA, Lundström A, Kelsell DP., Am J Hum Genet. 2013
</p>

<p>
A missense mutation in the MBTPS2 gene underlies the X-linked form of Olmsted syndrome. Haghighi A, Scott CA, Poon DS, Yaghoobi R, Saleh-Gohari N, Plagnol V, Kelsell DP. J Invest Dermatol. 2013
</p>

<p>
RHBDF2 mutations are associated with tylosis, a familial esophageal cancer syndrome.
Blaydon DC, Etheridge SL, Risk JM, Hennies HC, Gay LJ, Carroll R, Plagnol V, McRonald FE, Stevens HP, Spurr NK, Bishop DT, Ellis A, Jankowski J, Field JK, Leigh IM, South AP, Kelsell DP. Am J Hum Genet. 2012
</p>

<p>
Inflammatory skin and bowel disease linked to ADAM17 deletion.
Blaydon DC, Biancheri P, Di WL, Plagnol V, Cabral RM, Brooke MA, van Heel DA, Ruschendorf F, Toynbee M, Walne A, O'Toole EA, Martin JE, Lindley K, Vulliamy T, Abrams DJ, MacDonald TT, Harper JI, Kelsell DP. N Engl J Med. 2011
</p>

<b>Bone marrow failure:</b>

<p>
Constitutional mutations in RTEL1 cause severe dyskeratosis congenita.
Walne AJ, Vulliamy T, Kirwan M, Plagnol V, Dokal I. Am J Hum Genet. 2013
</p>

<p>
Exome sequencing identifies autosomal-dominant SRP72 mutations associated with familial aplasia and myelodysplasia. Kirwan M, Walne AJ, Plagnol V, Velangi M, Ho A, Hossain U, Vulliamy T, Dokal I. Am J Hum Genet. 2012
</p>

<p>
Exome sequencing identifies MPL as a causative gene in familial aplastic anemia.
Walne AJ, Dokal A, Plagnol V, Beswick R, Kirwan M, de la Fuente J, Vulliamy T, Dokal I.
Haematologica. 2012
</p>

<b>Cardiovascular disorders:</b>

<p>
Genetic complexity in hypertrophic cardiomyopathy revealed by high-throughput sequencing. Lopes LR, Zekavati A, Syrris P, Hubank M, Giambartolomei C, Dalageorgou C, Jenkins S, McKenna W; Uk10k Consortium, Plagnol V, Elliott PM. J Med Genet. 2013
</p>

<b>Neurology:</b>

<p>
Mutations in ANO3 cause dominant craniocervical dystonia: ion channel implicated in pathogenesis. Charlesworth G, Plagnol V, Holmström KM, Bras J, Sheerin UM, Preza E, Rubio-Agusti I, Ryten M, Schneider SA, Stamelou M, Trabzuni D, Abramov AY, Bhatia KP, Wood NW.
Am J Hum Genet. 2012
</p>

<p>
Kohlschütter-Tönz syndrome: mutations in ROGDI and evidence of genetic heterogeneity. Tucci A, Kara E, Schossig A, Wolf NI, Plagnol V, Fawcett K, Paisán-Ruiz C, Moore M, Hernandez D, Musumeci S, Tennison M, Hennekam R, Palmeri S, Malandrini A, Raskin S, Donnai D, Hennig C, Tzschach A, Hordijk R, Bast T, Wimmer K, Lo CN, Shorvon S, Mefford H, Eichler EE, Hall R, Hayes I, Hardy J, Singleton A, Zschocke J, Houlden H.
 Hum Mutat. 2013
 </p>

 <b>Other disease fields:</b>

 <p>
 Recessive oligodontia linked to a homozygous loss-of-function mutation in the SMOC2 gene, Alfawaz S, Fong F, Plagnol V, Wong FS, Fearne J, Kelsell DP. Arch Oral Biol. 2013 May;58(5):462-6. doi: 10.1016/j.archoralbio.2012.12.008. Epub 2013 Jan 11.
 </p>

 <p>
Cryptogenic multifocal ulcerating stenosing enteritis associated with homozygous deletion mutations in cytosolic phospholipase A2-α Brooke MA, Longhurst HJ, Plagnol V, Kirkby NS, Mitchell JA, Rüschendorf F, Warner TD, Kelsell DP, Macdonald TT.  Gut. 2012
</p>
  </div>


<div id="tab5" class="tab-pane">

   <b> The UCLex agreement </b>
<p>
Rare disease genetics has been revolutionized by the widespread use of exome sequencing. However, the interpretation of sequence data on a large scale raises two major issues. The first issue is technical, related to the limits of short read sequencing, and the artifacts that these tools can generate. The second limitation is the difficulty to interpret rare variants without the knowledge of the genetic variation in large cohorts of individuals.
</p>

<p>
 These issues can be addressed to a large extent by sharing of clinical sequence data across multiple groups. This allows the identification of shared traits and variants, and the verification of hypotheses by comparing clinical data. Another advantage is the improvement in calling accuracy when the same variants are detected across multiple samples.
 </p>

 <p>
 To this end, we have assembled the UCL-exomes (or UCLex) consortium. UCLex groups together a dozen clinical investigators (most at UCL but not only) with an analytical lead based at the UGI. Rules are in place to facilitate data sharing while respecting patient privacy. As of now, UCLex consists of {{total_patients}} exome samples and this set grows regularly. Data are stored on the UCL computer science cluster and analysis, including a monthly joint calling of the entire set, is centralized at the UCL Genetics Consortium. Owing to an average of 10 Gb size per exome, these sequence data add up to {{total_patients / 100}} Tb of required storage. This analysis creates significant computational challenges, and technical solutions (hardware/storage) are being developed as the sample size grows to accommodate this analysis. A key tool is the reduced reads format developed by the <a href=https://software.broadinstitute.org/gatk/>GATK</a> team.
 </p>


<p>
UCLex is a collaboration between clinical groups and analysts to facilitate the analysis of the large amount of exome sequence data being generated. The difficulty is to strike a balance between data confidentiality, data ownership but also the advantages associated with sharing data across multiple groups to facilitate variant calling, case control analysis and generally speaking the dissection of both complex and Mendelian disorders.
</p>

<p>
Clinical PIs are typically the “owners” of a exome sequence data and analysts at UCL provide calls/association tests back to the clinical groups. The general rules of engagement are listed below:
</p>

<ul>
<li> Vincent Plagnol will keep a list of PIs up to date, and if others want to join he will approve it provided that it creates no conflict with other PIs.
If there is a conflict, he will check that the relevant PIs are happy with any addition to UCLex.

<li>If any conflict came up (ie related diseases) he will make sure that data are properly separated and no conflict comes up (note that he is not aware of any such issue as of now).
Collaborations down the road are obviously welcome.

<li> All the data will remain on the UCL Genetics Institute server and is only analyzed within this collaboration.
No raw data is shared without explicit approval.

<li> Within that set of analysts, it is appropriate to use the data in any way that is useful for case control type analysis (gene based, single variant based...).

</ul>


</div>

</div>

<script src="/static/bower_components/requirejs/require.js"></script>
<script src="/static/js/hpo.js"></script>
<script type="text/javascript">

requirejs.config({
    //By default load any module IDs from js/lib
    baseUrl: 'js',
    //except, if the module ID starts with "app",
    //load it from the js/app directory. paths
    //config is relative to the baseUrl, and
    //never includes a ".js" extension since
    //the paths config could be for a directory.
    paths: {
        d3: '/static/bower_components/d3/d3',
       // "dot-checker": '/static/bower_components/graphviz-d3-renderer/dist/dot-checker',
        "layout-worker": '/static/bower_components/graph-viz-d3-js/dist/layout-worker',
        worker: '/static/bower_components/requirejs-web-workers/src/worker',
        renderer: '/static/bower_components/graph-viz-d3-js/dist/renderer'
    }
});
var hpo_data = {{hpo_json|safe}};
//initial draw
draw_hpo(hpo_data['data']);
//breakdown
$('input:radio').change(function(){
    $('#hpo-dot').empty();
    draw_hpo(hpo_data['data']);
});
//remove relatedness?
$('#remove-related').on('click',function(){
    $('#hpo-dot').empty();
    draw_hpo(hpo_data['data']);
});
function convert_to_dot_overview(data,breakdown){
    var dot = 'digraph {\nsize="6,4";\n';//ratio=fill;\n';
    // set up pie colour
    var color = d3.scale.category20();
    // draw legend
    $('#hpo-legend').empty();
    // get legend data {key:count}
    var legendData = {};
    $.each(data['HP:0000001'][breakdown],function(k1,v1){
        legendData[k1] = 0;
        $.each(v1,function(k2,v2){
            legendData[k1] += d3.sum(d3.values(v2));
        });
    });
    function sortLegendKeys(a,b){
        var aV = legendData[a];
        var bV = legendData[b];
        return ((aV < bV) ? 1 : ((aV > bV) ? -1 : 0));
    }
    // sort legendData's key
    var legendKeys = Object.keys(legendData).sort(sortLegendKeys);
    // define x axis for legend
    var radius = d3.scale.linear()
        .domain([0, d3.max(d3.values(legendData),function(d){return Math.sqrt(d)})])
        .range([0,20]);
    var legendSvg = d3.select('#hpo-legend')
        .append('svg')
        .attr('width',800)
        .attr('height',function(){
            return 50 * Math.ceil(legendKeys.length / 10);
        });
    var legendG = legendSvg.selectAll('g')
        .data(legendKeys)
        .enter()
        .append('g')
        .attr('transform',function(d,i){
            var x = 20 + i%10*80;
            var y = Math.floor(i/10)*50 +25;
            return 'translate('+x+','+y+')';
        });
    legendG.append('title')
        .text(function(d){return legendData[d];});
    legendG.append('circle')
        .attr('r',function(d){return radius(Math.sqrt(legendData[d]));})
        .attr('fill', function(d){return color(d)});
    legendG.append('text')
        .style('font-size','9px')
        .attr('dy','0.25em')
        .text(function(d){return d;});
    // draw nodes
    $.each(data, function(key,d){
        d.width = 2*d.freq;
        d.label = d.name;
        if (d.width < 0.4){
            d.label = '';
        }
        var fillcolor = [];
        // get color propotion
        // get total first
        var total = 0;
        $.each(legendData,function(k1,v1){
            if (k1 in d[breakdown]){
                $.each(d[breakdown][k1],function(k2,v2){
                    total += d3.sum(d3.values(v2));
                });
            }
        });
        // then propotion
        $.each(legendKeys,function(i,e){
            var n = 0;
            if (e in d[breakdown]){
                $.each(d[breakdown][e],function(k2,v2){
                    n += d3.sum(d3.values(v2));
                });
            }
            var propotion = n/total;
            fillcolor.push(color(e)+';'+propotion);
        });

        dot = dot + '"' + key +  '" [style="wedged", fixedsize="true", fontsize="6", shape="doublecircle", width="' + d.width +'", fillcolor="' + fillcolor.join(':') +'", label="' + d.label +'", id="' + key + '", color="transparent"];\n';
        //dot = dot + '"' + d.id +  '" [style="wedged", fixedsize="true", fontsize="6", shape="circle", width="' + d.width +'", fillcolor="red;0.3:green;0.6:orange", label="' + d.label +'", id="' + d.id + '", color="transparent"];\n';
    });
    // add links
    $.each(data, function(key,value){
        if ('is_a' in value){
            $.each(value['is_a'], function(i, is_a){
                dot = dot + '"' + is_a + '" -> "' + key +'" [color=grey, lty="solid"];\n';
            });
        }
    });
    return dot + '}';
}
function draw_hpo(temp){
    // get mode
    var mode = 'related';
    if ($('#remove-related').is(':checked')){
        mode = 'unrelated';
    }
    // get breakdown
    var breakdown = $('input[name=optradio]:checked').val();
    var my_data = jQuery.extend(true,{}, temp[mode]);
    var returned = transform_overview(my_data,breakdown);
    var num_of_patients = returned['HP:0000001'].count;
    $('#num_of_patients').text(num_of_patients);
    var dot = convert_to_dot_overview(returned,breakdown);
    var callback = function (data){
        svg = d3.select('#hpo-dot').select('svg');
        all_g = svg.select('g').select('g').selectAll('g.node');
        all_g
            .on('mouseover', showTooltip)
            .on('mouseout', removeTooltip);
        $('.node').on('click',function(d) {
                // get the HPO id
            var this_node = $(this);
            var id = this_node.find('title').text();
            window.location.href = "/hpo/"+id;
        });
    }
    require(["renderer"],
        function (renderer) {
            // initialize svg stage
            zoomFunc = renderer.init({element:"#hpo-dot", extend:[0.1, 10]});
            // update stage with new dot source
            renderer.render({source:dot, raw:returned, labelAttributer: myLabelAttributer, callBack:callback});
    });


    //Show the tooltip on the hovered over slice
    function removeTooltip(d) {

        //Hide tooltip
        $('.popover').each(function() {
            $(this).remove();
        }); 

    }//function removeTooltip
    function showTooltip(d) {
        //Define and show the tooltip
        $(this).popover({
                placement: 'auto top',
                container: '#hpo-dot',
                trigger: 'manual',
                position: 'fixed',
                html : true,
                title: function() { return d.id; },
                content: function() {
                    // looking for corresponding data in my_data
                    var thisData = returned[d.id];
                    var content = '<b>Name:</b> '+thisData.name;
                    // get keys and sort them
                    var keyData = {};
                    $.each(thisData[breakdown],function(k1,v1){
                        keyData[k1] = 0;
                        $.each(v1,function(k2,v2){
                            keyData[k1] += d3.sum(d3.values(v2));
                        });
                    });
                    function sortLegendKeys(a,b){
                        var aV = keyData[a];
                        var bV = keyData[b];
                        return ((aV < bV) ? 1 : ((aV > bV) ? -1 : 0));
                    }
                    // sort legendData's key
                    var Keys = Object.keys(keyData).sort(sortLegendKeys);
                    $.each(Keys,function(i,e){
                        content = content + '<br><b>'+e+':</b> '+keyData[e];
                    });
                    return content; 
                }
        });
        $(this).popover('show');


    }//function showTooltip
    // custom labelAttributer
    var myLabelAttributer = function() {
        this
            .attr("x", function (d) {
                    return d.x;
                    })
            .attr('y', function(d,i){
                    var pData = d3.select(this.parentNode).datum().labels;
                    return -d.y - 7*(i - pData.length/2 + 1);
                    })
            .text(function (d) {
                    return d.text;
                    });

        this.each(function(d) {
                var self = d3.select(this);
                d.style.map(function(e) {
                    switch (e.key) {
                    case "stroke":
                    return {key: "color", value: e.value};
                    case "font-size":
                    return {key: e.key, value: e.value + "px"};
                    default:
                    return e;
                    }
                    }).forEach(function(e) {
                        self.style(e.key, e.value);
                        });
                });
    };
}

</script>

{% endblock %}




