<h5> The UCLex agreement </h5>
<ol>
    <li>Rare disease genetics has been revolutionized by the widespread use of exome sequencing. However, the interpretation of sequence data on a large scale raises two major issues. The first issue is technical, related to the limits of short read sequencing, and the artifacts that these tools can generate. The second limitation is the difficulty to interpret rare variants without the knowledge of the genetic variation in large cohorts of individuals.</li>
    <li>These issues can be addressed to a large extent by sharing of clinical sequence data across multiple groups. This allows the identification of shared traits and variants, and the verification of hypotheses by comparing clinical data. Another advantage is the improvement in calling accuracy when the same variants are detected across multiple samples.</li>
    <li>To this end, we have assembled the UCL-exomes (or UCLex) consortium. UCLex groups together a dozen clinical investigators (most at UCL but not only) with an analytical lead based at the UGI. Rules are in place to facilitate data sharing while respecting patient privacy. As of now, UCLex consists of {{total_patients}} exome samples and this set grows regularly. Data are stored on the UCL computer science cluster and analysis, including a monthly joint calling of the entire set, is centralized at the UCL Genetics Consortium. Owing to an average of 10 Gb size per exome, these sequence data add up to {{total_patients / 100}} Tb of required storage. This analysis creates significant computational challenges, and technical solutions (hardware/storage) are being developed as the sample size grows to accommodate this analysis. A key tool is the reduced reads format developed by the <a href=https://software.broadinstitute.org/gatk/>GATK</a> team.</li>
    <li>UCLex is a collaboration between clinical groups and analysts to facilitate the analysis of the large amount of exome sequence data being generated. The difficulty is to strike a balance between data confidentiality, data ownership but also the advantages associated with sharing data across multiple groups to facilitate variant calling, case control analysis and generally speaking the dissection of both complex and Mendelian disorders.</li>
    <li>Clinical PIs are typically the “owners” of a exome sequence data and analysts at UCL provide calls/association tests back to the clinical groups. The general rules of engagement are listed below:</li>
    <ul class="browser-default ">
        <li> Vincent Plagnol will keep a list of PIs up to date, and if others want to join he will approve it provided that it creates no conflict with other PIs. If there is a conflict, he will check that the relevant PIs are happy with any addition to UCLex.</li>
        <li>If any conflict came up (ie related diseases) he will make sure that data are properly separated and no conflict comes up (note that he is not aware of any such issue as of now). Collaborations down the road are obviously welcome.</li>
        <li> All the data will remain on the UCL Genetics Institute server and is only analyzed within this collaboration. No raw data is shared without explicit approval. </li>
        <li> Within that set of analysts, it is appropriate to use the data in any way that is useful for case control type analysis (gene based, single variant based...). </li>
    </ul>
</ol>
